DOCSP-30540: insert fundamental guide (#11)
mongodb/docs-rust: source/fundamentals/crud/write-operations/insert.txt

Insert Documents
================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to **insert** documents into a MongoDB
collection.

Before you can find, update, and delete any documents in MongoDB, you need
to insert them. You can insert documents by using the following methods:

- ``insert_one()`` to insert one document
- ``insert_many()`` to insert one or more documents

The _id Field
-------------

In MongoDB, each document *must* contain a unique ``_id`` field.
You can take the following approaches to manage the ``_id`` field:

- Manage this field yourself, ensuring that each ``_id`` value you set
  is unique. 
- Let the driver automatically generate unique ``ObjectId`` values. The
  driver generates a unique ``ObjectId`` value for each document if you do
  not specify the ``_id`` field when performing an insert.

Unless you provide strong guarantees for uniqueness, we recommend that
you let the driver automatically generate ``_id`` values for your
documents.

.. important:: Duplicate _id Values

   Duplicate ``_id`` values violate unique index constraints, causing

>    the operation to fail with a ``BulkWriteFailure``.

Reviewer: S: Specify which operation(s) fail. 

Also torn on whether we need the name of the exception here. If you think it's important, maybe change "fail with" to "throw a BulkWriteFailure exception" or something similar.

To learn more about the ``_id`` field, see :manual:`Unique Indexes
</core/index-unique/>` in the Server manual.

To learn more about document structure and rules, see
:manual:`Documents </core/document>` in the Server manual.

Insert a Document
-----------------

Use the ``insert_one()`` method to insert a single document into a
collection.

Upon successful insertion, the method returns an
``InsertOneResult`` instance that contains the ``_id`` of the inserted
document.

Example
~~~~~~~

> .. TODO decide if using structs in CRUD fundamentals

Reviewer: Q: Will this be decided later?

Author: yes

The following example uses the ``insert_one()`` method to insert a
document into the ``books`` collection:

.. io-code-block::

   .. input:: /includes/fundamentals/code-snippets/crud/insert.rs
      :start-after: begin-insert-one
      :end-before: end-insert-one
      :language: rust
      :dedent:

   .. output:: 
      :language: console
      :visible: false

      Inserted document with _id: ObjectId("...")

.. include:: /includes/fundamentals/automatic-creation.rst

Modify insert_one Behavior
~~~~~~~~~~~~~~~~~~~~~~~~~~

You can modify the behavior of the ``insert_one()`` method by
constructing and passing an optional ``InsertOneOptions`` struct. The

> following table describes the available options to set in

Reviewer: S:

```suggestion
following table describes the options available in
```

``InsertOneOptions``:

.. list-table::
   :header-rows: 1
   :stub-columns: 1
   :class: compatibility-large

   * - Option
     - Description

   * - ``bypass_document_validation``

>      - | If ``true``, allows the write to opt-out of document-level

Reviewer: I: verb + preposition doesn't need a hyphen, afaik

S:
```suggestion
     - | If ``true``, allows the write to opt out of document-level
```

>      - | If ``true``, allows the write to opt-out of document-level

Reviewer: I: Would be clearer and [more stylish](https://www.mongodb.com/docs/meta/style-guide/writing/use-effective-verbs/#don-t-humanize-inanimate-objects) to avoid having the connector opt out.

S: What controls the validation decision here?

Author: you can set document validation on a collection, so you could say the collection has the validation rules

Author: edited the wording, lmk what you think

         validation. To learn more about validation, see :manual:`Schema
         Validation </core/schema-validation>` in the Server manual.

       | Type: ``bool``
       | Default: ``false``

   * - ``write_concern``
     - | The write concern for the operation. To learn more about write
         concerns, see :manual:`Write Concern
         </reference/write-concern/>` in the Server manual.
       
       | Type: ``WriteConcern``

   * - ``comment``
     - | An arbitrary ``Bson`` value tagged to the operation to trace

>          the operation through the database profiler, ``currentOp``, and

Reviewer: I: What does "tagged to the operation" mean? Also repeats "operation"

S: Could the comment be "attached" or "tied" to the operation or something similar? Could also replace "trace the operation" to something like "trace its execution"

Author: what about just trace it?

>          logs. This option is only available on {+server+} versions 4.4

Reviewer: ```suggestion
         logs. This option is available only on {+server+} versions 4.4
```

         and later.
    
       | Type: ``Bson``

The following code shows how to construct an ``InsertOneOptions``
instance:

.. literalinclude:: /includes/fundamentals/code-snippets/crud/insert.rs
   :start-after: begin-one-options
   :end-before: end-one-options
   :language: rust
   :copyable:
   :dedent:

Insert Multiple Documents
-------------------------

Use the ``insert_many()`` method to insert multiple documents into a
collection.

Upon successful insertion, the method returns an
``InsertManyResult`` instance that contains the ``_id`` values of the
inserted documents.

Example
~~~~~~~

> .. TODO decide if using structs in CRUD fundamentals

Reviewer: Same question as above

The following example uses the ``insert_many()`` method to insert
multiple documents into the ``books`` collection:

.. io-code-block::

   .. input:: /includes/fundamentals/code-snippets/crud/insert.rs
      :start-after: begin-insert-many
      :end-before: end-insert-many
      :language: rust
      :dedent:

   .. output:: 
      :language: console
      :visible: false

      Inserted documents with _ids:
      ObjectId("...")
      ObjectId("...")
      ObjectId("...")

.. include:: /includes/fundamentals/automatic-creation.rst

Modify insert_many Behavior
~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can modify the behavior of the ``insert_many()`` method by

> constructing and passing an optional ``InsertManyOptions`` struct. The

Reviewer: S: Can we omit "optional" here? The language already makes it sound optional

```suggestion
constructing and passing an ``InsertManyOptions`` struct. The
```

Author: changed in both instances

> following table describes the available options to set in

Reviewer: ```suggestion
following table describes the options available in
```

``InsertManyOptions``:

.. list-table::
   :header-rows: 1
   :stub-columns: 1
   :class: compatibility-large

   * - Option
     - Description

   * - ``bypass_document_validation``

>      - | If ``true``, allows the write to opt-out of document-level

Reviewer: Same comment as above

         validation. To learn more about validation, see :manual:`Schema
         Validation </core/schema-validation>` in the Server manual.

       | Type: ``bool``
       | Default: ``false``

   * - ``ordered``

>      - | If ``true``, the operation retuns when an insert fails without performing the

Reviewer: ```suggestion
     - | If ``true``, the operation returns when an insert fails without performing the
```

>      - | If ``true``, the operation retuns when an insert fails without performing the
>          remaining writes. If ``false``, even if an insert fails, the

Reviewer: I: ambiguous meaning here. guessing it should be something like this:

```suggestion
     - | If ``true``, when an insert fails, the operation returns without inserting
        the remaining documents. If ``false``, even if an insert fails, the
```

         operation continues with the remaining writes. To learn more
         about ordered inserts, see the :ref:`Ordered Behavior
         <rust-ordered-behavior>` section of this guide.
       
       | Type: ``bool``
       | Default: ``true``

   * - ``write_concern``
     - | The write concern for the operation. To learn more about write
         concerns, see :manual:`Write Concern
         </reference/write-concern/>` in the Server manual.
       
       | Type: ``WriteConcern``

   * - ``comment``
     - | An arbitrary ``Bson`` value tagged to the operation to trace
         the operation through the database profiler, ``currentOp``, and

>          logs. This option is only available on {+server+} versions 4.4

Reviewer: Same comments as above

         and later.
    
       | Type: ``Bson``

The following code shows how to construct an ``InsertManyOptions``
instance:

.. literalinclude:: /includes/fundamentals/code-snippets/crud/insert.rs
   :start-after: begin-many-options
   :end-before: end-many-options
   :language: rust
   :copyable:
   :dedent:

.. _rust-ordered-behavior:

> Ordered Behavior

Reviewer: I like this example. Does it make sense to mention/link to it above, and/or change the name to "Ordered Behavior Example"?

~~~~~~~~~~~~~~~~

> Assume you want to insert the following documents:

Reviewer: S:

```suggestion
Assume you want to insert the following documents into the ``books`` collection:
```

.. code-block:: json
   :copyable: false

   { "_id": 1, "title": "Where the Wild Things Are" }
   { "_id": 2, "title": "The Very Hungry Caterpillar" }
   { "_id": 1, "title": "Blueberries for Sal" }
   { "_id": 3, "title": "Goodnight Moon" }

If you attempt to insert these documents with default ``InsertManyOptions``, a
``BulkWriteError`` occurs during the third insert because of the duplicate
``_id`` value, but the documents before the error-producing document still get
inserted into your collection.

If you specify ``ordered`` as ``false`` in your ``InsertManyOptions``
instance, every insert that does not cause a duplicate key error

> completes successfully.

Author: reworded some things, but mostly adhered to this structure

The following code shows how to perform an unordered write operation to
insert the preceding documents:

.. literalinclude:: /includes/fundamentals/code-snippets/crud/insert.rs
   :start-after: begin-unordered
   :end-before: end-unordered
   :emphasize-lines: 8-9
   :language: rust
   :copyable:
   :dedent:

Even though this operation results in a ``BulkWriteError``, you can
still find the non-error-producing documents in your collection:

.. code-block:: json
   :copyable: false

   { "_id": 1, "title": "Where the Wild Things Are" }
   { "_id": 2, "title": "The Very Hungry Caterpillar" }
   { "_id": 3, "title": "Goodnight Moon" }

Additional Information
----------------------

.. TODO For runnable examples of the insert operations, see the following usage
.. examples:

.. TODO To learn more about performing the operations mentioned, see the
.. following guides:
.. 

> .. - :ref:`rust-query-guide`

Reviewer: Q: confirming these will be added later

Author: We will make a TODO resolution ticket at some point

     - Description

   * - ``bypass_document_validation``
     - | If ``true``, allows the driver to perform a write that violates
         document-level validation. To learn more about validation, see

>          :manual:`Schema Validation </core/schema-validation>` in the Server manual.

Reviewer: Really clear. nicely done

   { "_id": 1, "title": "Blueberries for Sal" }
   { "_id": 3, "title": "Goodnight Moon" }

When you attempt to insert these documents, the result depends on the
value of the ``ordered`` option in your ``InsertManyOptions``:

- If ``ordered`` is ``true`` (the default value), the driver throws a
  ``BulkWriteError`` when it attempts to insert the document with the
  duplicate ``_id`` value. However, the driver still inserts the
  documents before the error occurs. 

- If you set ``ordered`` to ``false``, the driver still throws a
  ``BulkWriteError`` when it attempts to insert the document with the

>   duplicate ``_id`` value, but it inserts every other document.

Reviewer: Looks good!
